### ISSUE TITLE
protoc-gen-grpc-java not available on apple m1

### ISSUE THREAD

**Issue by reinmind** [grpc/grpc-java#7690]
### What version of gRPC-Java are you using?
```1.34.0```
### What is your environment?

```osx 11.0.1```
### What did you expect to see?
```build success```
### What did you see instead?
```log
[INFO] --- protobuf-maven-plugin:0.6.1:compile (default-cli) @ grpc ---
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.720 s
[INFO] Finished at: 2020-12-03T16:37:02+08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.xolstice.maven.plugins:protobuf-maven-plugin:0.6.1:compile (default-cli) on project grpc: Unable to resolve artifact: Missing:
[ERROR] ----------
[ERROR] 1) com.google.protobuf:protoc:exe:osx-aarch_64:3.12.0
[ERROR] 
[ERROR]   Try downloading the file manually from the project website.
[ERROR] 
[ERROR]   Then, install it using the command: 
[ERROR]       mvn install:install-file -DgroupId=com.google.protobuf -DartifactId=protoc -Dversion=3.12.0 -Dclassifier=osx-aarch_64 -Dpackaging=exe -Dfile=/path/to/file
[ERROR] 
[ERROR]   Alternatively, if you host your own repository you can deploy the file there: 
[ERROR]       mvn deploy:deploy-file -DgroupId=com.google.protobuf -DartifactId=protoc -Dversion=3.12.0 -Dclassifier=osx-aarch_64 -Dpackaging=exe -Dfile=/path/to/file -Durl=[url] -DrepositoryId=[id]
[ERROR] 
[ERROR]   Path to dependency: 
[ERROR]   	1) com.reinmind:grpc:jar:1.0-SNAPSHOT
[ERROR]   	2) com.google.protobuf:protoc:exe:osx-aarch_64:3.12.0
[ERROR] 
[ERROR] ----------
[ERROR] 1 required artifact is missing.
[ERROR] 
[ERROR] for artifact: 
[ERROR]   com.reinmind:grpc:jar:1.0-SNAPSHOT
[ERROR] 
[ERROR] from the specified remote repositories:
[ERROR]   aliyunmaven (https://maven.aliyun.com/repository/public, releases=true, snapshots=false)
[ERROR] 
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

```

### Steps to reproduce the bug
```mvn protobuf:compile```

**Comment by ejona86** [comment]
Yes, we do not produce binaries for arm64 macs.

**Comment by ejona86** [comment]
Since this is the grpc repo, I've called out protoc-gen-grpc-java not being available. Although your actual error is _protoc_ itself, which would be part of the protobuf project. If grpc gets to building it first, then we'd need to add support in protobuf as well.

**Comment by gaurav46** [comment]
hi
Is it possible for any workaround like to set osdetector.arch as x86_64 to get the complication done ?

**Comment by ejona86** [comment]
To manually specify the classifier, for Gradle:
```
protobuf {
  protoc {
    artifact = 'com.google.protobuf:protoc:3.14.0:osx-x86_64'
  }
}
```

I'm less familiar with the Maven plugin, but I think it may be something like:
```
<protocArtifact>com.google.protobuf:protoc:3.14.0:exe:osx-x86_64</protocArtifact>
```

You can do the same for protoc-gen-grpc-java.

**Comment by gaurav46** [comment]
Thanks @ejona86 . That worked out.

**Comment by Jiachen-Zhang** [comment]
Thanks @ejona86 , But How can I keep other platform like linux-x86_64 when the project being compiled on other platform?

**Comment by ejona86** [comment]
@Jiachen-Zhang 

```
protobuf {
  protoc {
    if (osdetector.os == "osx") {
      artifact = 'com.google.protobuf:protoc:3.14.0:osx-x86_64'
    } else {
      artifact = 'com.google.protobuf:protoc:3.14.0'
    }
  }
}
```

You can also use project properties to only change the behavior when explicitly requested.

**Comment by Jiachen-Zhang** [comment]
> @Jiachen-Zhang
> 
> ```
> protobuf {
>   protoc {
>     if (osdetector.os == "osx") {
>       artifact = 'com.google.protobuf:protoc:3.14.0:osx-x86_64'
>     } else {
>       artifact = 'com.google.protobuf:protoc:3.14.0'
>     }
>   }
> }
> ```
> 
> You can also use project properties to only change the behavior when explicitly requested.

Thanks!

Finally, I made it by

```
// for apple m1, please add protoc_platform=osx-x86_64 in $HOME/.gradle/gradle.properties
if (project.hasProperty('protoc_platform')) {
    artifact = "com.google.protobuf:protoc:3.13.0:${protoc_platform}"
} else {
    artifact = "com.google.protobuf:protoc:3.13.0"
}
```

**Comment by guyeu** [comment]
for me, add this to `~/.m2/settings.xml`, it worked:

```xml
<settings>
  ...
  <activeProfiles>
    <activeProfile>
      apple-silicon
    </activeProfile >
    ...
  </activeProfiles>
  <profiles>
    <profile>
      <properties>
        <os.detected.classifier>osx-x86_64</os.detected.classifier>
      </properties>
    </profile>
    ...
  </profiles>
  ...
</settings>
```

**Comment by Uditmittal** [comment]
Can you share the full setting.xml 
I tried it didn't work for me?
I am using M1 with protobuf.

**Comment by Jiachen-Zhang** [comment]
@Uditmittal 
in my `build.gradle` file
```
protobuf {
    protoc {
        // for apple m1, please add protoc_platform=osx-x86_64 in $HOME/.gradle/gradle.properties
        if (project.hasProperty('protoc_platform')) {
            artifact = "com.google.protobuf:protoc:3.13.0:${protoc_platform}"
        } else {
            artifact = "com.google.protobuf:protoc:3.13.0"
        }
    }
}
```
in my `~/.gradle/gradle.properties` file

```
protoc_platform=osx-x86_64
```
I think you can define and use variables in maven to implement this feature. Good luck

**Comment by guyeu** [comment]
@Uditmittal My settings.xml is awfully long, the key is using your property overwrite `kr.motd.maven:os-maven-plugin` detected os classifier, if you share your settings.xml and pom.xml snippet, this may be help.

**Comment by Uditmittal** [comment]
<settings>
  <activeProfiles>
    <activeProfile>
      apple-silicon
    </activeProfile>
  </activeProfiles>
  <profiles>
    <profile>
      <id>apple-silicon</id>
      <properties>
        <os.detected.classifier>osx-x86_64</os.detected.classifier>
      </properties>
    </profile>
  </profiles>
</settings>

@guyeu my seetings.xml file

**Comment by guyeu** [comment]
@Uditmittal This is my `org.xolstice.maven.plugins:protobuf-maven-plugin` build plugin configuration, the `${os.detected.classifier}` value could read from active profile from `settings.xml`.

```xml
      <configuration>
          <protocArtifact>
              com.google.protobuf:protoc:3.1.0:exe:${os.detected.classifier}
          </protocArtifact>
          <pluginId>grpc-java</pluginId>
          <pluginArtifact>io.grpc:protoc-gen-grpc-java:1.0.0:exe:${os.detected.classifier}</pluginArtifact>
      </configuration>
```

**Comment by Uditmittal** [comment]
Thanks, it works for me.

**Comment by bangbang2333** [comment]
谢谢老哥，搞定了

**Comment by ejona86** [comment]
I'm not at all impressed with how hard it was to find the appropriate flags to pass to clang to make a universal binary, but [it appears to be `-arch arm64 -arch x86_64`](https://sqlite.org/forum/forumpost/f6831fd046). Unclear if cross-compiling is as simple as `-arch arm64`. We can start by figuring out how to cross-compile/universal binary for protobuf. Afterward we can fight Gradle. The goal would be to let us build M1 binaries from a aarch64 Mac. But I don't have a Mac which means I can't help much, as develop-via-CI is a form of torture.

**Comment by jjones-figure** [comment]
@ejona86 Have you had any success building from source for AArch64? I am interested in helping moving this along...

**Comment by ejona86** [comment]
@jjones-figure, no, as I said, I don't have a Mac, so I can't really help much here. We totally build from source for AArch64 already on Linux. But it is a different toolchain for M1.

**Comment by ejona86** [comment]
https://github.com/protocolbuffers/protobuf/pull/8557 just went into protobuf based on my suggestion to just copy the current x86 binary to an arm64 name until we have actual M1 support. That doesn't "fix" this, but it does resolve the main user-visible issues and doesn't turn out to be too hacky.

I think that is as simple as making a copy of the "exe" and the hashes/signature in our [upload_artifacts.sh script](https://github.com/grpc/grpc-java/blob/master/buildscripts/kokoro/upload_artifacts.sh).

**Comment by kdubb** [comment]
@ejona86 Where do I beg to get `protoc-gen-grpc-java` published as `osx-aarch_64`?

This is the lone holdout of 3 projects (including `com.google.protobuf:protoc`) required for building easily on the M1. The other projects seem to be following the advice of copying `osx-x86_64` to `osx-aarch_64`  and it works well. I currently do this manually on my machine to get the required `protoc-gen-grpc-java:osx-aarch_64` artifact.

**Comment by kdubb** [comment]
@ejona86 I've had a look at the repo and created a very simple PR #8680 . No idea if it works but maybe this gets things rolling?

**Comment by liuxuliangcumt** [comment]
> @Uditmittal in my `build.gradle` file
> 
> ```
> protobuf {
>     protoc {
>         // for apple m1, please add protoc_platform=osx-x86_64 in $HOME/.gradle/gradle.properties
>         if (project.hasProperty('protoc_platform')) {
>             artifact = "com.google.protobuf:protoc:3.13.0:${protoc_platform}"
>         } else {
>             artifact = "com.google.protobuf:protoc:3.13.0"
>         }
>     }
> }
> ```
> 
> in my `~/.gradle/gradle.properties` file
> 
> ```
> protoc_platform=osx-x86_64
> ```
> 
> I think you can define and use variables in maven to implement this feature. Good luck

in flutter_blue  android. build.gradle add protobuf...
  you say  in my `build.gradle` file .how could we know in' flutter_blue android  build.gradle 'to add ?

**Comment by solohsu** [comment]
https://gist.github.com/solohsu/a2bac0498ef51c105cedd2dc0a854bcf
I write a more general Gradle workaround script which may be helpful for you guys who come across any plugin that doesn't support macOS M1 Soc, not only protobuf.

**Comment by liuxuliangcumt** [comment]
谢谢您的来信！

**Comment by ilya-yarets** [comment]
> To manually specify the classifier, for Gradle:
> 
> ```
> protobuf {
>   protoc {
>     artifact = 'com.google.protobuf:protoc:3.14.0:osx-x86_64'
>   }
> }
> ```
> 
> I'm less familiar with the Maven plugin, but I think it may be something like:
> 
> ```
> <protocArtifact>com.google.protobuf:protoc:3.14.0:exe:osx-x86_64</protocArtifact>
> ```
> 
> You can do the same for protoc-gen-grpc-java.

thank you very much!

**Comment by ejona86** [comment]
Reopening because we've only a workaround currently. We still need to do something along the lines of https://github.com/grpc/grpc-java/issues/7690#issuecomment-831617279 for native M1 support. I _thought_ we had another issue open for tracking native M1 support, but I can't find it. Anyway, this will do.

**Comment by liuxuliangcumt** [comment]
谢谢您的来信！

**Comment by sukhbir-singh** [comment]
Is this solved ? I am still facing this issue.

**Comment by ejona86** [comment]
Interactions with the protobuf plugins is resolved, but requires Rosetta. I reopened the issue to track native M1 support.

**Comment by coding-devops** [comment]
> To manually specify the classifier, for Gradle:
> 
> ```
> protobuf {
>   protoc {
>     artifact = 'com.google.protobuf:protoc:3.14.0:osx-x86_64'
>   }
> }
> ```
> 
> I'm less familiar with the Maven plugin, but I think it may be something like:
> 
> ```
> <protocArtifact>com.google.protobuf:protoc:3.14.0:exe:osx-x86_64</protocArtifact>
> ```
> 
> You can do the same for protoc-gen-grpc-java.

why it's not work for me .... >_<

**Comment by mflis** [comment]
@ejona86  Is adding native M1 support dependent on resolving this [issue](https://github.com/protocolbuffers/protobuf/issues/9397#issuecomment-1086138036) first? 

I have Macbook M1 available, so I could experiment a bit, but I’m not very familiar with this repository, so I would appreciate pointing me in the right direction.

**Comment by ejona86** [comment]
@mflis, it is independent but related. If either us (grpc-wide, not just java) or protobuf confirm [how to crosscompile](https://github.com/grpc/grpc-java/issues/7690#issuecomment-831617279), then it will probably make the process a bit faster for the others. M1 macs are becoming a bit more common (the likelihood you know someone that has one), which is making this easier to test as time goes on.

**Comment by xizizzz** [comment]
[ERROR] Failed to execute goal org.xolstice.maven.plugins:protobuf-maven-plugin:0.5.0:compile (default) on project lollypop-proto: An error occurred while invoking protoc.: Error while executing process. Cannot run program "/Users/yinpeng/JavaWorkSpace/lollypop/lollypop-server/lollypop-proto/target/protoc-plugins/protoc-3.1.0-osx-x86_64.exe": error=86, Bad CPU type in executable -> [Help 1]
[ERROR] @guyeu maven 换完之后执行 还是报错

**Comment by xizizzz** [comment]
`[ERROR] Failed to execute goal org.xolstice.maven.plugins:protobuf-maven-plugin:0.5.0:compile (default) on project lollypop-proto: An error occurred while invoking protoc.: Error while executing process. Cannot run program "/Users/yinpeng/JavaWorkSpace/lollypop/lollypop-server/lollypop-proto/target/protoc-plugins/protoc-3.1.0-osx-x86_64.exe": error=86, Bad CPU type in executable -> [Help 1]
[ERROR] `
@guyeu maven 换完之后执行 还是报错

**Comment by amihaiemil** [comment]
For maven users, add this profile:

```xml
<profile>
    <id>m1</id>
    <build>
        <plugins>
            <plugin>
                <groupId>com.github.os72</groupId>
                 <artifactId>protoc-jar-maven-plugin</artifactId>
                 <executions>
                     <execution>
                         <phase>generate-sources</phase>
                         <goals>
                             <goal>run</goal>
                         </goals>
                         <configuration>
                             <protocArtifact>com.google.protobuf:protoc:3.1.0:exe:osx-x86_64</protocArtifact>
                         </configuration>
                     </execution>
                </executions>
            </plugin>
        </plugins>
    </build>
</profile>
```

The important thing is:

<protocArtifact>com.google.protobuf:protoc:3.1.0:exe:``osx-x86_64``</protocArtifact>

With osx-x86_64, the plugin knows to use the exe for Apple M1.

Then just ``mvn clean install -Pm1`` should work.

**Comment by amihaiemil** [comment]
@ejona86 Thanks, but what do you mean with ``(fake; copy of x86_64)``?

**Comment by ejona86** [comment]
@amihaiemil, see https://github.com/protocolbuffers/protobuf/pull/8557

**Comment by MochiXu** [comment]
@ejona86 Would you please tell me what's your Gradle version, my Gradle still try to download the aarch64 version after using your suggestion

**Comment by ejona86** [comment]
@MochiXu, which of my suggestions? And what error are you seeing?

**Comment by liuxuliangcumt** [comment]
谢谢您的来信！

**Comment by MochiXu** [comment]
> @MochiXu, which of my suggestions? And what error are you seeing?

your suggestions: https://github.com/grpc/grpc-java/issues/7690#issuecomment-747196215

**Comment by MochiXu** [comment]
> @MochiXu, which of my suggestions? And what error are you seeing?

Thanks for your reply! I see your suggestions from  this link:  https://github.com/grpc/grpc-java/issues/7690#issuecomment-747196215)，my computer use m1 chip, so when I try to use Gradle to build protobuf, Gradle will try to download protoc-3.14.0-osx-arrch_64.exe from https://repo1.maven.org/maven2/com/google/protobuf/protoc/3.14.0/protoc-3.14.0-osx-arrch_64.exe, but this repository doesn't have protoc-3.14.0-osx-arrch_64.exe. So under your suggestion, I changed artifact = 'com.google.protobuf:protoc:3.14.0' with artifact = 'com.google.protobuf:protoc:3.14.0:osx-x86_64', but it doesn't work, it is still downloading form https://repo1.maven.org/maven2/com/google/protobuf/protoc/3.14.0/protoc-3.14.0-osx-arrch_64.exe

**Comment by ejona86** [comment]
@MochiXu, I expect you're just missing a place in your build that you need to make the change. There's not much way to help you without seeing your build. That said, that recommendation of mine is pretty old. Since protobuf v3.17.3, protobuf has been copying the osx-x86_64 artifact to osx-aarch_64 meaning you wouldn't need configuration in your build. So these days I'd recommend you just upgrade to a newer protobuf.

**Comment by MochiXu** [comment]
> @MochiXu, I expect you're just missing a place in your build that you need to make the change. There's not much way to help you without seeing your build. That said, that recommendation of mine is pretty old. Since protobuf v3.17.3, protobuf has been copying the osx-x86_64 artifact to osx-aarch_64 meaning you wouldn't need configuration in your build. So these days I'd recommend you just upgrade to a newer protobuf.

It works!👍 Thanks for your kindness🙏, I changed my protobuf to v3.17.3 and grpc to v1.45.1✌️

**Comment by wallezhang** [comment]
@ejona86 Hi, I found the protoc-gen-grpc-java program is still not available on apple M1 chip. It seems the copy of x86 version, so I received an error message: program not found or is not executable when I compile the proto files.

**Comment by chenzhiguo** [comment]
Bump grpc-java.version to 1.47.0 and protobuf-java.version to 3.21.1

**Comment by wallezhang** [comment]
> Bump grpc-java.version to 1.47.0 and protobuf-java.version to 3.21.1 for Apple Silicon is ok.

Actually `protoc-gen-grpc-java` is not compatible with m1 chip since bump grpc-java version to 1.47.0. `protoc-gen-grpc-java-1.47.0-osx-aarch_64.exe` and `protoc-gen-grpc-java-1.47.0-osx-x86_64.exe` has the same md5, so there is no native compatible with m1 chip.

If you don't install Rosetta on mac, then you will receive an error when execute `protoc-gen-grpc-java-1.47.0-osx-aarch_64.exe`. And protobuf-java is compatible with m1 chip natively.

**Comment by Jalaja-Apexon** [comment]
Hi,
I am facing this issue while building my grpc project in Java using Maven build

[ERROR] Failed to execute goal com.github.os72:protoc-jar-maven-plugin:3.9.2:run (default) on project grpc-Oscar-Health: Error resolving artifact: com.google.protobuf:protobuf-java:jar:3.9.2:exe:osx-x86_64: Failure to find com.google.protobuf:protobuf-java:3.9.2:exe:jar in https://repo.maven.apache.org/maven2 was cached in the local repository, resolution will not be reattempted until the update interval of central has elapsed or updates are forced
[ERROR] 
[ERROR] Try downloading the file manually from the project website.
[ERROR] 
[ERROR] Then, install it using the command: 
[ERROR]     mvn install:install-file -DgroupId=com.google.protobuf -DartifactId=protobuf-java -Dversion=jar -Dclassifier=exe -Dpackaging=3.9.2 -Dfile=/path/to/file
[ERROR] 
[ERROR] Alternatively, if you host your own repository you can deploy the file there: 
[ERROR]     mvn deploy:deploy-file -DgroupId=com.google.protobuf -DartifactId=protobuf-java -Dversion=jar -Dclassifier=exe -Dpackaging=3.9.2 -Dfile=/path/to/file -Durl=[url] -DrepositoryId=[id]
[ERROR] 
[ERROR] 
[ERROR]   com.google.protobuf:protobuf-java:3.9.2:jar
[ERROR] 
[ERROR] from the specified remote repositories:
[ERROR]   central (https://repo.maven.apache.org/maven2, releases=true, snapshots=false)
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

My pom.xml file is as follows: - 

<?xml version="1.0" encoding="UTF-8"?>

<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <groupId>grpc-Oscar-Health</groupId>
  <artifactId>grpc-Oscar-Health</artifactId>
  <version>0.0.1-SNAPSHOT</version>
  <packaging>jar</packaging>

  <name>grpc-Oscar-Health</name>
  <dependencies>
	<dependency>
			<groupId>com.google.protobuf</groupId>
			<artifactId>protobuf-java</artifactId>
			<version>3.9.2</version>
		</dependency>
		
		<dependency>
   			 <groupId>io.grpc</groupId>
    		<artifactId>grpc-all</artifactId>
   			 <version>1.47.0</version>
		</dependency>
		
		<dependency>
			<groupId>io.grpc</groupId>
			<artifactId>grpc-netty-shaded</artifactId>
			<version>1.22.1</version>
		</dependency>
		<dependency>
			<groupId>io.grpc</groupId>
			<artifactId>grpc-protobuf</artifactId>
			<version>1.22.1</version>
		</dependency>
		<dependency>
			<groupId>io.grpc</groupId>
			<artifactId>grpc-stub</artifactId>
			<version>1.22.1</version>
		</dependency>
		
		<dependency>
		    <groupId>javax.annotation</groupId>
		    <artifactId>javax.annotation-api</artifactId>
		    <version>1.3.2</version>
		</dependency>
		
		<dependency>
        <groupId>org.slf4j</groupId>
        <artifactId>slf4j-simple</artifactId>
        <version>1.7.28</version>
    </dependency>
    		
	</dependencies>
	
	
		<properties>
		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
		<maven.compiler.source>11</maven.compiler.source>
    	<maven.compiler.target>11</maven.compiler.target>
	</properties>
	
	<build>
		<defaultGoal>clean generate-sources compile install</defaultGoal>

		<plugins>
			
			<plugin>
				<groupId>com.github.os72</groupId>
				<artifactId>protoc-jar-maven-plugin</artifactId>
				<version>3.9.2</version>
				<executions>
					<execution>
						<phase>generate-sources</phase>
						<goals>
							<goal>run</goal>
						</goals>
				<configuration>
					<protocVersion>3.9.2</protocVersion>
					<includeStdTypes>true</includeStdTypes>
					<protocArtifact>com.google.protobuf:protobuf-java:jar:3.9.2:exe:osx-x86_64</protocArtifact>
					<pluginArtifact>io.grpc:protoc-gen-grpc-java:pom:1.47.0</pluginArtifact>
							<includeMavenTypes>direct</includeMavenTypes>
							
							<includeMavenTypes>transitive</includeMavenTypes>
													
							<addProtoSources>all</addProtoSources>
						   
							<includeDirectories>
								<include>src/main/resources</include>
							</includeDirectories>
								<inputDirectories>
								<include>src/main/resources/protobuf/src</include>
							</inputDirectories>
							<inputDirectories>
								<include>src/main/resources/protobuf/src/hioscar</include>
							</inputDirectories>														
							<outputTargets>
								<outputTarget>
									<type>java</type>
									<outputDirectory>src/main/java</outputDirectory>
								</outputTarget>
							</outputTargets>
						</configuration>
					</execution>
				</executions>
			</plugin>


			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-compiler-plugin</artifactId>
				<version>3.8.0</version>
				<configuration>
					<source>1.8</source>
					<target>1.8</target>
				</configuration>
			</plugin>
		</plugins>
	</build>
 
</project>

Kindly suggest suitable resolution for the same. The entire setup is in Mac device.

**Comment by ejona86** [comment]
@Jalaja-Apexon, you are not suffering from this issue. This is for protoc-gen-grpc-java and your error is about protocArtifact. Although both of them are defined improperly.

```
<protocArtifact>com.google.protobuf:protobuf-java:jar:3.9.2:exe:osx-x86_64</protocArtifact>
<pluginArtifact>io.grpc:protoc-gen-grpc-java:pom:1.47.0</pluginArtifact>
```

The `:jar` and `:pom` stuff shouldn't be there, it should be 'exe' after the version. You're referencing protobuf-java instead of protoc. You probably want to upgrade to a newer version of protoc which supports M1 natively. Follow https://github.com/grpc/grpc-java#generated-code

**Comment by Jalaja-Apexon** [comment]
Sure.. 
I have followed the pom.xml structure defined in the link - https://github.com/grpc/grpc-java#generated-code. However, should I also include the below tag as well
<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-compiler-plugin</artifactId>
				<version>3.8.0</version>
				<configuration>
					<source>1.8</source>
					<target>1.8</target>
				</configuration>
			</plugin>

**Comment by Jalaja-Apexon** [comment]
<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-compiler-plugin</artifactId>
				<version>3.8.0</version>
				<configuration>
					<source>1.8</source>
					<target>1.8</target>
				</configuration>
			</plugin>

**Comment by Jalaja-Apexon** [comment]
And how about <includeDirectories> or <inputDirectories>? Where should these tags be following the pom.xml

**Comment by ejona86** [comment]
@Jalaja-Apexon, your problem is off-topic for this issue. Please create a new issue.

**Comment by wallezhang** [comment]
@ejona86 So is there a plan to support M1 chip natively?

**Comment by joschi** [comment]
@wallezhang This probably depends on https://github.com/protocolbuffers/protobuf/issues/9397.

**Comment by wallezhang** [comment]
> @wallezhang This probably depends on [protocolbuffers/protobuf#9397](https://github.com/protocolbuffers/protobuf/issues/9397).

I see last comment in protocolbuffers/protobuf/issues/9397 said that OSX aarch64 binary has been already included.

**Comment by joschi** [comment]
@wallezhang https://github.com/protocolbuffers/protobuf/issues/9397#issuecomment-1124400911 also says:

> Unfortunately, we have not been able sign it yet due to internal infrastructure problems. I will continue to keep you updated on that when there is progress.

**Comment by slachiewicz** [comment]
please adjust plugin configuration to something like this
```
<plugins>
            <!-- compile proto file into java files. -->
            <plugin>
                <groupId>com.github.os72</groupId>
                <artifactId>protoc-jar-maven-plugin</artifactId>
                <version>3.11.4</version>
                <executions>
                    <execution>
                        <phase>generate-sources</phase>
                        <goals>
                            <goal>run</goal>
                        </goals>
                        <configuration>
                            <protocVersion>3.21.1</protocVersion>
                            <includeStdTypes>true</includeStdTypes>
                            <pluginArtifact>io.grpc:protoc-gen-grpc-java:pom:1.47.0</pluginArtifact>
                            <includeMavenTypes>direct</includeMavenTypes>
                            <includeMavenTypes>transitive</includeMavenTypes>
                            <addProtoSources>all</addProtoSources>
                            <includeDirectories>
                                <include>src/main/resources</include>
                            </includeDirectories>
                            <inputDirectories>
                                <include>src/main/resources/protobuf/src</include>
                            </inputDirectories>
                            <outputTargets>
                                <outputTarget>
                                    <type>java</type>
                                    <outputDirectory>src/main/java</outputDirectory>
                                </outputTarget>
                            </outputTargets>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
```

my output
```
[INFO] ----------------< grpc-Oscar-Health:grpc-Oscar-Health >-----------------
[INFO] Building grpc-Oscar-Health 0.0.1-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- protoc-jar-maven-plugin:3.11.4:run (default) @ grpc-Oscar-Health ---
[INFO] Protoc version: 3.21.1
protoc-jar: protoc version: 3.21.1, detected platform: osx-aarch_64 (mac os x/aarch64)
protoc-jar: cached: /var/folders/46/zh5s3qfd0j7gr6x4ffj03s_h0000gn/T/protocjar.webcache/com/google/protobuf/protoc/maven-metadata.xml
protoc-jar: cached: /var/folders/46/zh5s3qfd0j7gr6x4ffj03s_h0000gn/T/protocjar.webcache/com/google/protobuf/protoc/3.21.1/protoc-3.21.1-osx-aarch_64.exe
protoc-jar: executing: [/var/folders/46/zh5s3qfd0j7gr6x4ffj03s_h0000gn/T/protocjar2051368796002136723/bin/protoc.exe, --version]
libprotoc 3.21.1
[INFO] Protoc command: /var/folders/46/zh5s3qfd0j7gr6x4ffj03s_h0000gn/T/protocjar2051368796002136723/bin/protoc.exe
[INFO] Additional include types: /var/folders/46/zh5s3qfd0j7gr6x4ffj03s_h0000gn/T/protocjar13749638867177729208/include
[INFO]     google/protobuf/any.proto
[INFO]     google/protobuf/api.proto
[INFO]     google/protobuf/compiler/plugin.proto
[INFO]     google/protobuf/descriptor.proto
[INFO]     google/protobuf/duration.proto
[INFO]     google/protobuf/empty.proto
[INFO]     google/protobuf/field_mask.proto
[INFO]     google/protobuf/source_context.proto
[INFO]     google/protobuf/struct.proto
[INFO]     google/protobuf/timestamp.proto
[INFO]     google/protobuf/type.proto
[INFO]     google/protobuf/wrappers.proto
[INFO]     grpc/lb/v1/load_balancer.proto
[INFO]     grpc/lookup/v1/rls_config.proto
[INFO]     grpc/lookup/v1/rls.proto
[INFO]     grpc/reflection/v1alpha/reflection.proto
[INFO]     grpc/health/v1/health.proto
[INFO]     grpc/channelz/v1/channelz.proto
[INFO]     grpc/binlog/v1/binarylog.proto
[INFO]     grpc/gcp/altscontext.proto
[INFO]     grpc/gcp/transport_security_common.proto
[INFO]     grpc/gcp/handshaker.proto
[INFO]     opencensus/proto/trace/v1/trace_config.proto
[INFO]     opencensus/proto/trace/v1/trace.proto
[INFO]     opencensus/proto/resource/v1/resource.proto
[INFO]     opencensus/proto/metrics/v1/metrics.proto
[INFO]     opencensus/proto/stats/v1/stats.proto
[INFO]     opencensus/proto/agent/trace/v1/trace_service.proto
[INFO]     opencensus/proto/agent/common/v1/common.proto
[INFO]     opencensus/proto/agent/metrics/v1/metrics_service.proto
[INFO]     google/api/monitoring.proto
[INFO]     google/api/distribution.proto
[INFO]     google/api/backend.proto
[INFO]     google/api/annotations.proto
[INFO]     google/api/label.proto
[INFO]     google/api/billing.proto
[INFO]     google/api/monitored_resource.proto
[INFO]     google/api/config_change.proto
[INFO]     google/api/consumer.proto
[INFO]     google/api/context.proto
[INFO]     google/api/experimental/experimental.proto
[INFO]     google/api/experimental/authorization_config.proto
[INFO]     google/api/control.proto
[INFO]     google/api/metric.proto
[INFO]     google/api/system_parameter.proto
[INFO]     google/api/usage.proto
[INFO]     google/api/httpbody.proto
[INFO]     google/api/http.proto
[INFO]     google/api/auth.proto
[INFO]     google/api/source_info.proto
[INFO]     google/api/endpoint.proto
[INFO]     google/api/logging.proto
[INFO]     google/api/service.proto
[INFO]     google/api/log.proto
[INFO]     google/api/documentation.proto
[INFO]     google/api/quota.proto
[INFO]     google/longrunning/operations.proto
[INFO]     google/cloud/audit/audit_log.proto
[INFO]     google/logging/type/log_severity.proto
[INFO]     google/logging/type/http_request.proto
[INFO]     google/type/date.proto
[INFO]     google/type/latlng.proto
[INFO]     google/type/timeofday.proto
[INFO]     google/type/color.proto
[INFO]     google/type/money.proto
[INFO]     google/type/dayofweek.proto
[INFO]     google/type/postal_address.proto
[INFO]     google/rpc/code.proto
[INFO]     google/rpc/status.proto
[INFO]     google/rpc/error_details.proto
[INFO] Input directories:
[INFO]     /Users/slachiewicz/work/grpc-java-7690/src/main/resources/protobuf/src
[INFO] Include directories:
[INFO]     /Users/slachiewicz/work/grpc-java-7690/src/main/resources
[INFO]     /var/folders/46/zh5s3qfd0j7gr6x4ffj03s_h0000gn/T/protocjar13749638867177729208/include
[INFO] Output targets:
[INFO]     java: /Users/slachiewicz/work/grpc-java-7690/src/main/java (add: main, clean: false, plugin: null, outputOptions: null)
[INFO]     Processing (java): book.proto
protoc-jar: executing: [/var/folders/46/zh5s3qfd0j7gr6x4ffj03s_h0000gn/T/protocjar2051368796002136723/bin/protoc.exe, -I/Users/slachiewicz/work/grpc-java-7690/src/main/resources, -I/var/folders/46/zh5s3qfd0j7gr6x4ffj03s_h0000gn/T/protocjar13749638867177729208/include, -I/Users/slachiewicz/work/grpc-java-7690/src/main/resources/protobuf/src, --java_out=/Users/slachiewicz/work/grpc-java-7690/src/main/java, /Users/slachiewicz/work/grpc-java-7690/src/main/resources/protobuf/src/book.proto]
[INFO] Adding generated sources (java): /Users/slachiewicz/work/grpc-java-7690/src/main/java
```

**Comment by krebznet** [comment]
Thank You For This!

> for me, add this to `~/.m2/settings.xml`, it worked:
> 
> ```
> <settings>
>   ...
>   <activeProfiles>
>     <activeProfile>
>       apple-silicon
>     </activeProfile>
>     ...
>   </activeProfiles>
>   <profiles>
>     <profile>
>       <id>apple-silicon</id>
>       <properties>
>         <os.detected.classifier>osx-x86_64</os.detected.classifier>
>       </properties>
>     </profile>
>     ...
>   </profiles>
>   ...
> </settings>
> ```

**Comment by tisonkun** [comment]
When debugging https://github.com/apache/pulsar/issues/16639, I found that protoc 3.21.2 provides a native (real) osx-aarch_64 binary, while protoc-gen-java-plugin leaves as an alias to the x86_64 version (even in 1.47.0). Is there any plan to provides a native osx-aarch_64 binary of protoc-gen-java-plugin?

**Comment by ravileeladhar2023** [comment]
> ```
> <settings>
>   <activeProfiles>
>     <activeProfile>
>       apple-silicon
>     </activeProfile>
>   </activeProfiles>
>   <profiles>
>     <profile>
>       <id>apple-silicon</id>
>       <properties>
>         <os.detected.classifier>osx-x86_64</os.detected.classifier>
>       </properties>
>     </profile>
>   </profiles>
> </settings>
> ```
> 
> 
>      
>   
> 
> @guyeu my seetings.xml file

Hi  Can you please share the complete settings.xml I have added above config but still I am getting below error 

[ERROR] Failed to execute goal com.github.os72:protoc-jar-maven-plugin:3.11.4:run (generate-sources) on project fcpsinboundadapter: Error extracting protoc for version 3.11.4: Unsupported platform: protoc-3.11.4-osx-aarch_64.exe -> [Help 1]

**Comment by liuxuliangcumt** [comment]
谢谢您的来信！

**Comment by ravileeladhar2023** [comment]
> @Uditmittal This is my `org.xolstice.maven.plugins:protobuf-maven-plugin` build plugin configuration, the `${os.detected.classifier}` value could read from active profile from `settings.xml`.
> 
> ```
>       <configuration>
>           <protocArtifact>
>               com.google.protobuf:protoc:3.1.0:exe:${os.detected.classifier}
>           </protocArtifact>
>           <pluginId>grpc-java</pluginId>
>           <pluginArtifact>io.grpc:protoc-gen-grpc-java:1.0.0:exe:${os.detected.classifier}</pluginArtifact>
>       </configuration>
> ```

Can you explain , I couldn't able to find the plugin org.xolstice.maven.plugins:protobuf-maven-plugin and how to add this configuration and where

**Comment by tangcent** [comment]
I update my `build.gradle`  looks like this:
```gradle
protobuf {
    protoc {
        if (System.getProperty("os.arch") == "aarch64") {
            // mac
            artifact = "com.google.protobuf:protoc:3.10.0:osx-x86_64"
        } else {
            // linux
            artifact = "com.google.protobuf:protoc:3.10.0"
        }
    }
}
```

**Comment by dtretyakov** [comment]
@ejona86 do you know when the `protoc-gen-grpc-java_xxx_osx_aarch64` build of will be fixed? As of now `osx_aarch` and `osx_x86-84` are the same binary which is not compatible with M1 CPU:
```
./protoc-gen-grpc-java-1.51.1-osx-aarch_64.exe 
zsh: bad CPU type in executable: ./protoc-gen-grpc-java-1.51.1-osx-aarch_64.exe
```
Source: https://repo1.maven.org/maven2/io/grpc/protoc-gen-grpc-java/1.51.1/

I suspect it's not big deal for infra team to make a proper build for that artifact since `protoc-xxx-osx-aarch_64.exe` was built correctly: https://repo1.maven.org/maven2/com/google/protobuf/protoc/3.21.9/

Am I right that @deannagarcia could do that? https://github.com/protocolbuffers/protobuf/issues/9397#issuecomment-1124400911

If it's impossible to provide correct `aarch64` builds of `protoc-gen-grpc-java` please remove it from artifact publishing to not mislead end users.

**Comment by liuxuliangcumt** [comment]
谢谢您的来信！

**Comment by ejona86** [comment]
@dtretyakov, the binary can work via Rosetta. Yes, it is a lie, but it makes using the Maven and Gradle plugins much easier and with Rosetta installed the lie isn't really noticable.

The only hold-up is for someone to get the cross-compile working.
https://github.com/grpc/grpc-java/issues/7690#issuecomment-831617279

**Comment by Hadrien-Cornier** [comment]
@hengyunabc  's answer worked for me !

> ```
>     <profiles>
>         <profile>
>             <id>mac</id>
>             <activation>
>                 <os> 
>                 <family>mac</family>
>                 </os>
>             </activation>
>             <properties>
>                 <os.detected.classifier>osx-x86_64</os.detected.classifier>
>             </properties>
>         </profile>
>     </profiles>
> ```

**Comment by ejona86** [comment]
> hengyunabc 's answer worked for me !

If that worked for you, then you have Rosetta installed and just need to upgrade to grpc 1.43 or later (#8680). This is issue is remaining open for native M1 builds, to avoid the Rosetta requirement.

**Comment by Hadrien-Cornier** [comment]
> If that worked for you, then you have Rosetta installed and just need to upgrade to grpc 1.43 or later (https://github.com/grpc/grpc-java/pull/8680). This is issue is remaining open for native M1 builds, to avoid the Rosetta requirement.
Thank you, I have a MacOS Ventura 13.3.1 with an M1 chip. I think hengyunabc's answer worked [because of Rosetta, as you said](https://www.imore.com/macos-ventura-includes-support-rosetta-arm-linux-vms)

**Comment by joschi** [comment]
https://github.com/protocolbuffers/protobuf/issues/9397#issuecomment-1535044692 🥳

**Comment by liuxuliangcumt** [comment]
谢谢您的来信！

**Comment by venshine** [comment]
`/usr/sbin/softwareupdate --install-rosetta --agree-to-license`
It is now May 2023 and none of the above methods work. This command solved my problem.

**Comment by mickamy** [comment]
updating yer software is yer problem, and not of this oss. it is yer obligation.On May 5, 2023, at 18:59, venshine ***@***.***> wrote:﻿
/usr/sbin/softwareupdate --install-rosetta --agree-to-license
It is now May 2023 and none of the above methods work. This command solved my problem.

**Comment by IgorZaporozhtsev** [comment]
> To manually specify the classifier, for Gradle:
> 
> ```
> protobuf {
>   protoc {
>     artifact = 'com.google.protobuf:protoc:3.14.0:osx-x86_64'
>   }
> }
> ```
> 
> I'm less familiar with the Maven plugin, but I think it may be something like:
> 
> ```
> <protocArtifact>com.google.protobuf:protoc:3.14.0:exe:osx-x86_64</protocArtifact>
> ```
> 
> You can do the same for protoc-gen-grpc-java.

```
protobuf{
    protoc {
        protoc {
            artifact = 'com.google.protobuf:protoc:3.14.0:osx-x86_64'
        }
        plugins{
            grpc {
                artifact = 'io.grpc:grpc-netty-shaded:1.56.0'
            }
        }


    }
    // default proto plugin generate stub in build folder
    // change the stub generate folder
    //generatedFilesBaseDir = "$projectDir/src/generated"
}

```

```
sourceSets{
    main{
        proto{
            srcDirs 'src/main/java/com/zeecoder/springbootdemo/proto'
        }
        java{
            srcDirs 'build/generated/source/proto/main/grpc'
            srcDirs 'build/generated/source/proto/main/java'
        }
    }
}
```
<img width="350" alt="image" src="https://github.com/grpc/grpc-java/assets/35780695/a4445b3d-c8a8-47bf-afb0-229c05f93f1b">

Only one Java class was generated

**Comment by liuxuliangcumt** [comment]
谢谢您的来信！

**Comment by glli** [comment]
It's easy to build the native protoc-gen-grpc-java-$VER-osx-aarch64.exe with Bazel on a macbook m1/m2, even you can backfill the old versions for your project's backward compatibility, e.g. v1.31.1 ($VER=1.31.1)
1. Download https://github.com/grpc/grpc-java/archive/refs/tags/v1.31.1.tar.gz and extract it
2. Download https://github.com/bazelbuild/bazel/releases/download/6.3.2/bazel-6.3.2-darwin-arm64, rename it to bazel and make it executable
3. cd grpc-java-1.31.1/compiler and run `bazel build grpc_java_plugin`
4. cd grpc-java-1.31.1/bazel-bin/compiler and cp/mv grpc_java_plugin to ~/bin (or what ever folder you like)
5. make a symbolic link `ln -s ~/bin/grpc_java_plugin ~/.gradle/caches/modules-2/files-2.1/io.grpc/protoc-gen-grpc-java/1.31.1/$SHA1/protoc-gen-grpc-java-1.31.1-osx-aarch_64.exe`

(I didn't try building higher versions, please reply if you meet any issues in the above steps)

I also tried building grpc_java_plugin with Gradle. No problem of running `./gradlew :grpc-compiler:compileJava_pluginExecutableJava_pluginCpp` (need to compile protobuf lib first based on the COMPILING.md). However I was stuck at running `./gradlew :grpc-compiler:build`, always got
```
Task :grpc-compiler:linkJava_pluginExecutable FAILED
Undefined symbols for architecture arm64:
  "std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::at(unsigned long) const", referenced from:
      google::protobuf::io::Tokenizer::IsIdentifier(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) in libprotobuf.a(tokenizer.o)
```
Tried configuring `--host=arm64` when compiling protobuf and it's not recognized. I think probably there's discrepancy between arm64 and aarch64 at the lib linking step.

**Comment by liuxuliangcumt** [comment]
谢谢您的来信！

**Comment by JackyAnn** [comment]
zsh: bad CPU type in executable: ./protoc-gen-grpc-java-osx-aarch_64.exe

can not executable on mac m2 chips
version:
[protoc-gen-grpc-java-1.58.0-osx-aarch_64.exe](https://repo1.maven.org/maven2/io/grpc/protoc-gen-grpc-java/1.58.0/protoc-gen-grpc-java-1.58.0-osx-aarch_64.exe)

**Comment by liuxuliangcumt** [comment]
谢谢您的来信！

**Comment by Kishan10008** [comment]
> for me, add this to `~/.m2/settings.xml`, it worked:
> 
> ```
> <settings>
>   ...
>   <activeProfiles>
>     <activeProfile>
>       apple-silicon
>     </activeProfile>
>     ...
>   </activeProfiles>
>   <profiles>
>     <profile>
>       <id>apple-silicon</id>
>       <properties>
>         <os.detected.classifier>osx-x86_64</os.detected.classifier>
>       </properties>
>     </profile>
>     ...
>   </profiles>
>   ...
> </settings>
> ```

where i can find this .m2

**Comment by Kishan10008** [comment]
> zsh: bad CPU type in executable: ./protoc-gen-grpc-java-osx-aarch_64.exe
> 
> can not executable on mac m2 chips version: [protoc-gen-grpc-java-1.58.0-osx-aarch_64.exe](https://repo1.maven.org/maven2/io/grpc/protoc-gen-grpc-java/1.58.0/protoc-gen-grpc-java-1.58.0-osx-aarch_64.exe)

similar issue i am facing

**Comment by ejona86** [comment]
@JackyAnn, it should work on M2 as well, as long as you have Rosetta installed.

**Comment by gelald** [comment]
@hu-chia simple、useful、effective!

**Comment by liuxuliangcumt** [comment]
谢谢您的来信！

**Comment by hakusai22** [comment]
```xml
  <properties>
    <os.detected.classifier>osx-x86_64</os.detected.classifier>
  </properties>
```

**Comment by kdubb** [comment]
This is going to have to be fixed. Apple has not shipped x86 processors for a year. Going by their past switch, from Power -> x86, it will only be a couple years before they stop shipping Rosetta.

**Comment by isarang** [comment]
```
<plugin>
        <groupId>org.xolstice.maven.plugins</groupId>
        <artifactId>protobuf-maven-plugin</artifactId>
        <version>0.5.1</version>
        <configuration>
          <protocArtifact>com.google.protobuf:protoc:3.24.4:exe:${os.detected.classifier}</protocArtifact>
          <pluginId>grpc-java</pluginId>
          <pluginArtifact>io.grpc:protoc-gen-grpc-java:1.43.3:exe:${os.detected.classifier}</pluginArtifact>
        </configuration>
        <executions>
          <execution>
            <goals>
              <goal>compile</goal>
              <goal>compile-custom</goal>
            </goals>
          </execution>
        </executions>
      </plugin>
```
This works for me

**Comment by vvWantHealthLife** [comment]
thanks that's work

**Comment by martin-traverse** [comment]
March 2024. My build started working when I installed Rosetta. I'm using the Gradle protobuf plugin and did not need to change the Gradle build scripts. I echo the comments above, it would be good to have native binaries before the drop-dead date.

Another note, it seems this is already done for the protoc compiler itself, only the gRPC plugin is causing problems. I have three modules using protoc, only one of which also uses the gRPC plugin. The others just use base protoc and they built fine without needing to install Rosetta. Can the same approach that was used for the base protoc be replicated here?

**Comment by ejona86** [comment]
@martin-traverse, protobuf's release build used to be like ours. But they changed it dramatically and it is now based on Bazel. [Looks like they are using `--cpu=`](https://github.com/protocolbuffers/protobuf/blob/016a4b56fe3a826de03eb97db9063aebf9b5c271/toolchain/toolchains.bazelrc#L13-L14) to build aarch_64 and x86_64 individually. They have quite specific [toolchain configuration](https://github.com/protocolbuffers/protobuf/blob/016a4b56fe3a826de03eb97db9063aebf9b5c271/toolchain/BUILD.bazel#L97). On the release, they have three osx binaries: aarch_64, x86_64, and universal_binary. I don't see the code that makes the universal binary.

We are using autotools right now to build protobuf, and then using Gradle to build the plugin. But we've not been able to absorb the absl dependency in newer protobuf versions. I have [a change to swap to cmake](https://github.com/grpc/grpc-java/pull/10048) to build protobuf, but pkg-config is now required and that caused trouble on Windows. More recently pkgconf has taken off, so we might now have a solution there.

Since absl was added to protobuf, I have considered moving the plugin compilation out of Gradle. That could be a stepping-stone to using Bazel. But I think that is too much to depend on here. I still believe this just needs a few flags when building (like https://github.com/grpc/grpc-java/issues/7690#issuecomment-831617279, or the `--cpu` flags). But that really needs someone with a Mac to figure it out.

**Comment by Areeb786123** [comment]
> @Uditmittal in my `build.gradle` file
> 
> ```
> protobuf {
>     protoc {
>         // for apple m1, please add protoc_platform=osx-x86_64 in $HOME/.gradle/gradle.properties
>         if (project.hasProperty('protoc_platform')) {
>             artifact = "com.google.protobuf:protoc:3.13.0:${protoc_platform}"
>         } else {
>             artifact = "com.google.protobuf:protoc:3.13.0"
>         }
>     }
> }
> ```
> 
> in my `~/.gradle/gradle.properties` file
> 
> ```
> protoc_platform=osx-x86_64
> ```
> 
> I think you can define and use variables in maven to implement this feature. Good luck



> @Jiachen-Zhang
> 
> ```
> protobuf {
>   protoc {
>     if (osdetector.os == "osx") {
>       artifact = 'com.google.protobuf:protoc:3.14.0:osx-x86_64'
>     } else {
>       artifact = 'com.google.protobuf:protoc:3.14.0'
>     }
>   }
> }
> ```
> 
> You can also use project properties to only change the behavior when explicitly requested.

it didn't work for me

**Comment by gaofeiseu** [comment]
in my case, this will works well:
if you have a pom file and use maven plugin to complie the target package, for example:
![image](https://github.com/grpc/grpc-java/assets/5344460/dcc8961d-d781-44bd-949d-4c061761a7a6)
so, what you need to do is specify a fixed value for the position pointed by the arrow, the fixed value is osx-x86_64 , it's x86 mac os actual value.

**Comment by siddhsql** [comment]
what is the purpose of `protoc-gen-grpc-java` when the actual compiler is [protoc](https://github.com/protocolbuffers/protobuf/releases/tag/v26.1)?

**Comment by siddhsql** [comment]
any update on this issue? everyone is on M1/M2 chips these days.

**Comment by amraupward** [comment]
> any update on this issue? everyone is on M1/M2 chips these days.

M2 Macbook Air. 

build.gradle

protobuf {
    protoc {
        artifact = "com.google.protobuf:protoc:3.25.1"
    }
    plugins {
        grpc {
            artifact = 'io.grpc:protoc-gen-grpc-java:1.52.1'      // <----- this version can build your java 
        }
    }
    generateProtoTasks {
        all()*.plugins {
            grpc {}
        }
    }
}

if you want to create docker image. 

FROM jdk as gradle_cache
RUN mkdir -p /tmp/gradle/cache
ENV GRADLE_USER_HOME /tmp/gradle/cache
ARG GPR_USERNAME
ARG GPR_TOKEN
ENV GPR_USERNAME $GPR_USERNAME
ENV GPR_TOKEN $GPR_TOKEN
ENV GRADLE_OPTS="-Xmx4g -XX:+UseContainerSupport -Dfile.encoding=UTF-8".   // <----- this is important 
COPY build.gradle /tmp
WORKDIR /tmp/
RUN gradle clean build -i --stacktrace --no-daemon

and 

dockerbuild.sh
docker buildx build \
  --platform linux/amd64 --load \              //<------------ this too.
  -t "$LOCAL_TAG" \
  -t "$AZURE_LOCAL_TAG" \
  --build-arg GPR_USERNAME=$GPR_USERNAME \
  --build-arg GPR_TOKEN=$GPR_TOKEN \
  --quiet=false \
  --output=type=docker \
  -f ./docker.dockerfile .

**Comment by Karm** [comment]
Oh, it seems I hit this again, trying to download osx-aarch64 file only to find out it's actually a Mach-0 x86_64 binary:  https://github.com/grpc/grpc.io/issues/1296

**Comment by Ivan-Kurochenko** [comment]
```gradle
protobuf {
    protoc {
        if (osdetector.os == "osx") {
            artifact = 'com.google.protobuf:protoc:3.14.0:osx-x86_64'
        } else {
            artifact = 'com.google.protobuf:protoc:3.14.0'
        }
    }
}
```


I have pasted this code, and I have received another error. 


```
Execution failed for task ':generateProto'.
> Could not resolve all files for configuration ':protobufToolsLocator_grpc'.
   > Could not find protoc-gen-grpc-java-1.28.1-osx-aarch_64.exe (io.grpc:protoc-gen-grpc-java:1.28.1).
     Searched in the following locations:
         https://repo.maven.apache.org/maven2/io/grpc/protoc-gen-grpc-java/1.28.1/protoc-gen-grpc-java-1.28.1-osx-aarch_64.exe

* Try:
```

**Comment by Karm** [comment]
@Ivan-Kurochenko There is no such file on that location, is it? See https://repo.maven.apache.org/maven2/io/grpc/protoc-gen-grpc-java/1.28.1/

If you upgrade to e.g. https://repo.maven.apache.org/maven2/io/grpc/protoc-gen-grpc-java/1.64.0/ you can see there is an osx-aarch_64 file now.

The problem is though that it is in fact x86_64 binary and your aarch64 system won't run it:

```
% file ./protoc-gen-grpc-java-1.64.0-osx-aarch_64.exe
./protoc-gen-grpc-java-1.64.0-osx-aarch_64.exe: Mach-O 64-bit executable x86_64
```
```
zsh: bad CPU type in executable: ./protoc-gen-grpc-java-1.64.0-osx-aarch_64.exe
```

So you'd need to have Rosetta installed so as your Mac OS can translate amd64 machine code do aarch64 machine code...

**Comment by Ivan-Kurochenko** [comment]
@Karm I installed Rosetta. 
```
% lsbom -f /Library/Apple/System/Library/Receipts/com.apple.pkg.RosettaUpdateAuto.bom
./Library/Apple/usr/lib/libRosettaAot.dylib	100755	0/0	320896	219430798
./Library/Apple/usr/libexec/oah/RosettaLinux/rosetta	100755	0/0	16608881085004749
./Library/Apple/usr/libexec/oah/RosettaLinux/rosettad	100755	0/0	298792	794611390
./Library/Apple/usr/libexec/oah/libRosettaRuntime	100755	0/0	381680	959649284
./Library/Apple/usr/share/rosetta/rosetta	100644	0/0	64	1875722922

```

**Comment by Ivan-Kurochenko** [comment]
@Karm Do you have another idea, of how I can get rid of the protoc.3.13 for arm?

**Comment by Karm** [comment]
@Ivan-Kurochenko I am sorry, I don't understand the question.

**Comment by huydqdev** [comment]
i dont use Rosetta and this one work for me
```
protobuf {
  protoc {
    artifact = 'com.google.protobuf:protoc:3.14.0'
    path = '/opt/homebrew/bin/protoc'
  }
}
```
after 
`brew install protobuf`
and find the path
`which protoc`

**Comment by yash-brahmkshatriya** [comment]
For M1 chip,

I had to install rosetta to make it work.

`
    




    <properties>
        <protoc.version>4.27.1</protoc.version>
        <grpc.version>1.64.0</grpc.version>

        <maven.compiler.source>17</maven.compiler.source>
        <maven.compiler.target>17</maven.compiler.target>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    </properties>


    <dependencies>
        <dependency>
            <groupId>com.google.protobuf</groupId>
            <artifactId>protobuf-java</artifactId>
            <version>${protoc.version}</version>
        </dependency>
        <dependency>
            <groupId>com.google.protobuf</groupId>
            <artifactId>protobuf-java-util</artifactId>
            <version>${protoc.version}</version>
        </dependency>
        <dependency>
            <groupId>io.grpc</groupId>
            <artifactId>protoc-gen-grpc-java</artifactId>
            <version>${grpc.version}</version>
            <type>pom</type>
        </dependency>
        <dependency>
            <groupId>io.grpc</groupId>
            <artifactId>grpc-stub</artifactId>
            <version>${grpc.version}</version>
        </dependency>
        <dependency>
            <groupId>io.grpc</groupId>
            <artifactId>grpc-protobuf</artifactId>
            <version>${grpc.version}</version>
        </dependency>
        <dependency>
            <groupId>org.apache.tomcat</groupId>
            <artifactId>annotations-api</artifactId>
            <version>6.0.53</version>
            <scope>provided</scope>
        </dependency>
    </dependencies>

    <build>
        <extensions>
            <extension>
                <groupId>kr.motd.maven</groupId>
                <artifactId>os-maven-plugin</artifactId>
                <version>1.7.1</version>
            </extension>
        </extensions>
        <plugins>
            <plugin>
                <groupId>org.xolstice.maven.plugins</groupId>
                <artifactId>protobuf-maven-plugin</artifactId>
                <version>0.6.1</version>
                <configuration>
                    <protocArtifact>com.google.protobuf:protoc:${protoc.version}:exe:${os.detected.classifier}</protocArtifact>
                    <pluginId>grpc-java</pluginId>
                    <pluginArtifact>io.grpc:protoc-gen-grpc-java:${grpc.version}:exe:${os.detected.classifier}</pluginArtifact>
                </configuration>
                <executions>
                    <execution>
                        <goals>
                            <goal>compile</goal>
                            <goal>compile-custom</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
        </plugins>
    </build>`

**Comment by shalk** [comment]
i read https://github.com/grpc/grpc-java/releases/tag/v1.62.2
> The protoc plugin no longer supports macOS Big Sur (macOS 11). Binaries are now built using Monterey (macOS 12)

What is the change ?

**Comment by gharia** [comment]
Is there any solution with our hack to this? It works adding specific value as mentioned above, `os.detected.classifier` through settings.xml, but that is not a feasible solution when different team members using different OS.

**Comment by ejona86** [comment]
@gharia, you don't need to do any os.detected.classifier hacks. Just make sure Rosetta is installed on OS X arm machines and use releases from 2022 or newer.

**Comment by gharia** [comment]
I have Rosetta installed an proton version is 27.1

```
hemraj@192 .m2 % if /usr/bin/pgrep -q oahd; then echo 'rosetta installed'; fi
rosetta installed
hemraj@192 .m2 % protoc --version                                            
libprotoc 27.1
```

**Comment by gharia** [comment]
@ejona86 , could you help with this? Let me know if I am missing anything here

**Comment by shalk** [comment]
> For M1 chip,
> 
> I had to install rosetta to make it work.
> 
> ```
>     <properties>
>         <protoc.version>4.27.1</protoc.version>
>         <grpc.version>1.64.0</grpc.version>
> 
>         <maven.compiler.source>17</maven.compiler.source>
>         <maven.compiler.target>17</maven.compiler.target>
>         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
>     </properties>
> 
> 
>     <dependencies>
>         <dependency>
>             <groupId>com.google.protobuf</groupId>
>             <artifactId>protobuf-java</artifactId>
>             <version>${protoc.version}</version>
>         </dependency>
>         <dependency>
>             <groupId>com.google.protobuf</groupId>
>             <artifactId>protobuf-java-util</artifactId>
>             <version>${protoc.version}</version>
>         </dependency>
>         <dependency>
>             <groupId>io.grpc</groupId>
>             <artifactId>protoc-gen-grpc-java</artifactId>
>             <version>${grpc.version}</version>
>             <type>pom</type>
>         </dependency>
>         <dependency>
>             <groupId>io.grpc</groupId>
>             <artifactId>grpc-stub</artifactId>
>             <version>${grpc.version}</version>
>         </dependency>
>         <dependency>
>             <groupId>io.grpc</groupId>
>             <artifactId>grpc-protobuf</artifactId>
>             <version>${grpc.version}</version>
>         </dependency>
>         <dependency>
>             <groupId>org.apache.tomcat</groupId>
>             <artifactId>annotations-api</artifactId>
>             <version>6.0.53</version>
>             <scope>provided</scope>
>         </dependency>
>     </dependencies>
> 
>     <build>
>         <extensions>
>             <extension>
>                 <groupId>kr.motd.maven</groupId>
>                 <artifactId>os-maven-plugin</artifactId>
>                 <version>1.7.1</version>
>             </extension>
>         </extensions>
>         <plugins>
>             <plugin>
>                 <groupId>org.xolstice.maven.plugins</groupId>
>                 <artifactId>protobuf-maven-plugin</artifactId>
>                 <version>0.6.1</version>
>                 <configuration>
>                     <protocArtifact>com.google.protobuf:protoc:${protoc.version}:exe:${os.detected.classifier}</protocArtifact>
>                     <pluginId>grpc-java</pluginId>
>                     <pluginArtifact>io.grpc:protoc-gen-grpc-java:${grpc.version}:exe:${os.detected.classifier}</pluginArtifact>
>                 </configuration>
>                 <executions>
>                     <execution>
>                         <goals>
>                             <goal>compile</goal>
>                             <goal>compile-custom</goal>
>                         </goals>
>                     </execution>
>                 </executions>
>             </plugin>
>         </plugins>
>     </build>
> ```

I notice the protobuf 4.x have been supported . It is great!

**Comment by gharia** [comment]
Thanks @shalk those changes worked for. However I am using Protobuf version 3.25.3. I am on M2 chip

**Comment by ejona86** [comment]
@gharia, you don't say what problem you are experiencing. What error do you get?

Our examples work on Rosetta-available OS X ARM64 systems. Try the [Quick Start](https://grpc.io/docs/languages/java/quickstart/). It uses [configuration essentially the same](https://github.com/grpc/grpc-java/blob/master/examples/pom.xml) as shalk posted, but with protobuf 3.25.3. It's also essentially the same configuration as our main README.md.

@shalk, protobuf 4 still has an issue: https://github.com/protocolbuffers/protobuf/issues/17247

**Comment by gharia** [comment]
@ejona86 , I was using older version of Protobuf release (before 2022 as you mentioned) in pom.xml

Below is my working pom.xml

```
<protobuf.version>3.25.3</protobuf.version>
<io.grpc.version>1.65.1</io.grpc.version>
<os72.version>3.11.4</os72.version>

<build>
        <plugins>
            <plugin>
                <groupId>com.github.os72</groupId>
                <artifactId>protoc-jar-maven-plugin</artifactId>
                <version>${os72.version}</version>
                <executions>
                    <execution>
                        <phase>generate-sources</phase>
                        <goals>
                            <goal>run</goal>
                        </goals>
                        <configuration>
                            <protocArtifact>com.google.protobuf:protoc:${protobuf.version}</protocArtifact>
                            <inputDirectories>
                                <include>src/main/resources/protobuf</include>
                            </inputDirectories>
                            <outputTargets>
                                <outputTarget>
                                    <type>java</type>
                                </outputTarget>
                                <outputTarget>
                                    <type>grpc-java</type>
                                    <pluginArtifact>io.grpc:protoc-gen-grpc-java:${io.grpc.version}</pluginArtifact>
                                </outputTarget>
                            </outputTargets>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
        </plugins>
</build>			
```

**Comment by Mosiki** [comment]
Execute this command in terminal to install Rosetta 2:
`/usr/sbin/softwareupdate --install-rosetta`
Type `A` and press Enter to confirm

Specify pom as osx-x86_64 type, reference:
```xml
            <plugin>
                <groupId>org.xolstice.maven.plugins</groupId>
                <artifactId>protobuf-maven-plugin</artifactId>
                <version>0.6.1</version>
                <configuration>
                    <protocArtifact>
                            com.google.protobuf:protoc:3.5.1:exe:osx-x86_64
                    </protocArtifact>
                    <pluginId>grpc-java</pluginId>
                    <pluginArtifact>
                            io.grpc:protoc-gen-grpc-java:1.13.1:exe:osx-x86_64
                    </pluginArtifact>
                    <protoSourceRoot>src/main/proto</protoSourceRoot>
                </configuration>
                <executions>
                    <execution>
                        <goals>
                            <goal>compile</goal>
                            <goal>compile-custom</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
```
Hope this helps!

**Comment by mikehelmick** [comment]
FWIW - I was able to get this working on an M3 w/ arm64 architecture. The only change I made was explicitly setting JAVA_HOME to a JDK version of 11 or higher. Setting to JDK 8, it does indeed require the x86_64

**Comment by chenrui333** [comment]
FYI, homebrew added protoc-gen-grpc-java formula recently and it works for the arm build for apache-pulsar, https://github.com/Homebrew/homebrew-core/pull/199494

**Comment by slankka** [comment]
**~~They already support Apple Silicon~~**

**~~protoc-gen-java-grpc and~~ `protoc` plugin have Apple ARM versions.**

1) protoc
Since protobuf 3.17.3, we do not need change `os.detected.classifier` to `osx-x86-64`


2) protoc-gen-java-grpc 
Since protoc-gen-java-grpc 1.43.0,  ~~they also support Apple Silicon. See ref links below(release note of v1.43.0).~~

for protoc
**No need to install rosetta**, only need to upgrade protoc artifact version at least`3.17.3` 
 
for protoc-gen-java-grpc
upgrade protoc-gen-java-grpc at least `1.43.0+`. update your maven pom.xml version.

Check: protoc-gen-java-grpc rename `osx-x86_x64` to `osx-aarch_64`
**https://github.com/grpc/grpc-java/pull/8680/files** 

Verify for truth:
```
protoc-plugins % file protoc-3.25.5-osx-aarch_64.exe 
protoc-3.25.5-osx-aarch_64.exe: Mach-O 64-bit executable arm64
protoc-plugins % file protoc-gen-grpc-java-1.43.0-osx-aarch_64.exe 
protoc-gen-grpc-java-1.43.0-osx-aarch_64.exe: Mach-O 64-bit executable x86_64
```


example: 

```com.google.protobuf:protoc:3.25.6:exe:${os.detected.classifier}```
```io.grpc:protoc-gen-grpc-java:1.43.0:exe:${os.detected.classifier}```

**classifier is important**
note that the maven `classifier` `${os.detected.classifiler}` will be `osx-aarch_64`. 

**if we do not set classifier, maven will download default arch we may not want.**

**Compatibility with Hadoop**
WARN：Please be aware of that since **version 3.25.6 has breaking changes**, it will conflict with Hadoop ecosystems.

If your system calls hadoop rpc api, 3.25.6 fails with a runtime exception, see drill pr below for details. So, use 3.25.5 instead.

[protocolbuffers/protobuf#8695](https://github.com/protocolbuffers/protobuf/issues/8695)
https://github.com/apache/drill/pull/2977#issue-2812547525
https://github.com/grpc/grpc-java/releases/tag/v1.43.0
https://github.com/grpc/grpc-java/pull/8680

**Comment by evil0th** [comment]
I DO NOT like rosetta, i found the solution and it works:

1. install from brew

`brew install protoc-gen-grpc-java@your_version` 

2. link file

`ln -s /opt/homebrew/bin/protoc-gen-grpc-java /path/your/project/target/protoc-plugins/protoc-gen-grpc-java-your_version-osx-aarch_64.exe`

**Comment by bklebe** [comment]
@ejona86 thanks for the quick reply! It appears that Homebrew has aarch64-native binaries, but I assume they must be compiling them _on_ aarch64 rather than cross-compiling. For now, I guess the best course is to symlink the Homebrew artifact into the Gradle path? That's what I'm doing at least.

**Comment by ejona86** [comment]
In Gradle you should be able to replace `artifact` with `path` and just give it a path to where the binary is located or maybe a plain binary name but with the binary in PATH.

In Maven... for the xolstice plugin I clearly see `protocExecutable` as an alternative to `protocArtifact`, but it appears there isn't the equivalent for plugins. For the ascopes plugin, I see `binaryPathPlugins`, which lets you have the binary in PATH.

**Comment by waterlx** [comment]
Hi @kdubb @ejona86 

I was also blocked by this issue when I was trying to build Spark project.

So regarding your summary
> [#7690 (comment)](https://github.com/grpc/grpc-java/issues/7690#issuecomment-738179026) is still current. [#8680](https://github.com/grpc/grpc-java/pull/8680) purposefully copies the amd64 build to aarch64, as a stop-gap until someone gets cross compilation working. See also [google/protobuf-gradle-plugin#492](https://github.com/google/protobuf-gradle-plugin/issues/492) .

(as well as some other discussions in this thread) There are 2 options for M-series Apple Silicon CPU, right?

1. Install Rosetta.
	a. By `softwareupdate --install-rosetta`
	b. The reason behind that is [PR 8680](https://github.com/grpc/grpc-java/pull/8680), where OSX aarch_64 binaries are actually copies of OSX x64_64 (for Intel CPU). So Rosetta is needed to act as a translator for those binaries to run on M-series Apple Silicon CPU.
2. Install an OSX aarch_64 native binary
	a. By `brew install protoc-gen-grpc-java`
	b. In Maven, use `protocExecutable` to point to `protoc`, or set up a symbol link, like [this comment](https://github.com/grpc/grpc-java/issues/7690#issuecomment-2723767627)

**Comment by kdubb** [comment]
It's almost unfathomable to me that this hasn't been fixed yet. I created the PR to copy x86 and use Rosetta because I had no idea how to get it building otherwise as I was not (and still am not) really familiar with the project setup; that PR took me all of 5 minutes.

Is there no maintainer that knows the build well enough to actually build an aarch binary?

I could see if the, apparent, requirement of building it on macOS was an impediment, but it's clearly not since GitHub provides macOS runners for free for open source projects.

FYI, Apple is sunsetting Rosetta (as they did with PowerPC Rosetta), apparently macOS 27 is the last version. Normally I would think that is a long way off but considering everybody seems to be depending on my "5 minute fix" for the last **four** years... I'm not so sure.

If everybody's fine with kludgy fixes, somebody should at least make a PR to start a macOS runner, install the aarch binary via Homebrew, and copy the damn thing back during the packaging job.
